### PR TITLE
Replace GitHub with PagerDuty in our Incident Response process

### PR DESCRIPTION
**Pull request by yuvipanda**
PagerDuty is specifically tailored towards handling incidents, so let's use that rather
than try to rig an incident response process on top of GitHub ourselves. I also want us
to focus on the *Incident response team* primarily accessing PagerDuty via their slack
integration, rather than having to swap out to GitHub.

This PR only seeks to replace GitHub with PagerDuty / Slack, and makes no other
changes for now. 

A summary of the changes are:

1. New incidents are *triggered* by creating an Incident in PagerDuty
2. During an incident, the incident response team interacts with the incident
    primarily via slack, although they can do so via the web interface too
3. After the incident, we use the [pagerduty postmortems](https://support.pagerduty.com/docs/postmortems)
    feature to collect information and make a postmortem. This has several features that
    make this process much easier than editing comments on GitHub.

Ref https://github.com/2i2c-org/infrastructure/issues/1118

### CODE REVIEW

**Comment by yuvipanda** [comment]
Me and @jmunroe spent a bunch of time talking about this, and in particular role-playing how this *should* have played out with today's UToronto outage.

A proposed workflow is that after the incident is created, we want pagerduty to evaluate the current local timezone and stated preferences of all engineers and then send everyone who wants to be notified at that point a notification, via methods of their own choosing (SMS, App, Phone call). Engineers can then *acknowledge* the alert if they are able to provide assistance, *or do nothing* if they can't. After an hour (or a configurable time period later!), if nobody has acknowledged the alert, it'll automatically escalate to a (opt-in) group of second tier folks who can respond.

We're going to play with pagerduty rules to try make this possible.

However, even without *any* of this, it still provides value by streamlining the 'incident response' process and removing a medium (GitHub) that currently is a bit of a bottleneck - and instead moving it all to slack.

**Comment by yuvipanda** [comment]
This made me realize that I could reduce the scope of this PR to *just* replacing GitHub with Pagerduty - that alone would've caught that the incident from last week wasn't actually 'over', and would have prevented today's UToronto outage!

**Comment by yuvipanda** [comment]
https://github.com/2i2c-org/infrastructure/issues/1118#issuecomment-1240284103 has follow-up tasks

**Comment by yuvipanda** [comment]
I've tried to follow the incident response suggestions here, and created one for the UToronto outage yesterday https://2i2c-org.pagerduty.com/postmortems/171317d6-5f19-7511-7d3a-117b13f62584

File: projects/managed-hubs/incidents.md
Comment (sgibson91) [inline projects/managed-hubs/incidents.md:93]:
```suggestion
   information for the incident report after the fact.
```

File: projects/managed-hubs/incidents.md
Comment (sgibson91) [inline projects/managed-hubs/incidents.md:156]:
```suggestion
      `Opened the cloud console and discovered notifications about quota`. Pasting in commands is very helpful! This is an
```

File: projects/managed-hubs/incidents.md
Comment (sgibson91) [inline projects/managed-hubs/incidents.md:169]:
```suggestion
7. Click "Save & View Report" when you are done, and ask other members of the incident response team to review the incident report.
```

**Review by sgibson91 — APPROVED** [review]

File: projects/managed-hubs/incidents.md
Comment (choldgraf) [inline projects/managed-hubs/incidents.md:79]: We should have a team learning session for how to use the PagerDuty interface. Could somebody record themselves going through the full PagerDuty workflow as part of these instructions? (maybe we could use something like https://loom.com/ ?)

File: projects/managed-hubs/incidents.md
Comment (choldgraf) [inline projects/managed-hubs/incidents.md:61]: Ah we should also add a bullet-point for `incident-specific` channels since we create those below.

File: projects/managed-hubs/incidents.md
Comment (choldgraf) [inline projects/managed-hubs/incidents.md:93]: How do we define the External Liason? Before we said "use the issue" for this but I don't think there's a way to explicitly say this in PagerDuty

File: projects/managed-hubs/incidents.md
Comment (choldgraf) [inline projects/managed-hubs/incidents.md:105]: I moved all of this into a dedicated section so it didn't clutter these to-do lists too much.

File: projects/managed-hubs/incidents.md
Comment (choldgraf) [inline projects/managed-hubs/incidents.md:113]: When can the incident commander assign reporting duties to somebody else? I think if we don't define this then they will always be the ones that end up doing it. If we have this documented elsewhere we should cross-link it

File: projects/managed-hubs/incidents.md
Comment (choldgraf) [inline projects/managed-hubs/incidents.md:119]: This documentation suggests that this requires a paid plan. Is that true for us? If so we should make an explicit decision that we wish to pay for this: https://support.pagerduty.com/docs/postmortems

File: projects/managed-hubs/incidents.md
Comment (choldgraf) [inline projects/managed-hubs/incidents.md:146]: Signal-boosting this question from when I was making edits. What do you think about a Google Drive folder that exists just for the purpose of sharing incident reports?

If not that, maybe we could have a dedicated documentation site for this instead of using `infrastructure/`?

File: projects/managed-hubs/incidents.md
Comment (choldgraf) [inline projects/managed-hubs/incidents.md:154]: I think the easiest thing we can do is link to pre-existing timelines that are well-written. Then people can just riff off of the structure of those

**Review by choldgraf — COMMENTED** [review]
This seems like a really helpful service to use in order to provide some structure to our incident response process. In general, it looks good to me. I had a few comments and suggestions throughout.

One concern I have is that this is a fairly involved process. How can we ensure that we reliably and diligently follow this process?

Argh I took a pass and added some suggested edits to up the lists into different sections, but accidentally pushed to your branch instead of making a PR. Happy to discuss and I'll provide some comments below to focus discussion.

File: projects/managed-hubs/incidents.md
Comment (yuvipanda) [reply projects/managed-hubs/incidents.md:146]: I'm ok with it being in a different repo. Let's just do that, I don't want to use Google Drive for this - it's far less public.

File: projects/managed-hubs/incidents.md
Comment (yuvipanda) [reply projects/managed-hubs/incidents.md:155]: @choldgraf i agree! once we have a few more incidents we can incorporate that here.

File: projects/managed-hubs/incidents.md
Comment (yuvipanda) [reply projects/managed-hubs/incidents.md]: @choldgraf I have:

1. created https://github.com/2i2c-org/incident-reports
1. created https://github.com/2i2c-org/infrastructure/pull/1703 to move the existing 1 incident report we have in infrastructure repo to there
2. modified this PR to point to that.

File: projects/managed-hubs/incidents.md
Comment (yuvipanda) [reply projects/managed-hubs/incidents.md:119]: @choldgraf I opened https://github.com/2i2c-org/meta/issues/374

File: projects/managed-hubs/incidents.md
Comment (yuvipanda) [reply projects/managed-hubs/incidents.md:113]: @choldgraf tried to clarify this.

File: projects/managed-hubs/incidents.md
Comment (yuvipanda) [reply projects/managed-hubs/incidents.md]: @choldgraf I've updated with some suggestions

File: projects/managed-hubs/incidents.md
Comment (yuvipanda) [reply projects/managed-hubs/incidents.md]: Done, and reworded this a little

**Comment by yuvipanda** [comment]
Thanks a lot @sgibson91 and @choldgraf!

@choldgraf I think I've addressed all your concerns.

Re: learning to use pagerduty, there is a lot of documentation they maintain at https://university.pagerduty.com/ - would that be enough? I'm also happy to try make a video, but do not want to block this PR on that.

This process IMO is actually *simpler* than what we have now, because:

1.  *during* an incident, all interactions happen via slack (which we already use!)
2. Constructing a timeline becomes *much* easier - it's a fully manual process now
3. Easier to see when an incident can be marked as 'completed', a bit more difficult with GitHub now.

We will adjust this as we go along! However, I wrote this document so obviously I'll feel this way :D Would love to hear from others if they think this is too complex!

File: projects/managed-hubs/incidents.md
Comment (GeorgianaElena) [inline projects/managed-hubs/incidents.md:129]:
```suggestion
If they are *not* the person doing it, they should still create the incident report, but assign `Owner of the Review Process`
```

File: projects/managed-hubs/incidents.md
Comment (GeorgianaElena) [inline projects/managed-hubs/incidents.md:187]: ❤️

**Review by GeorgianaElena — APPROVED** [review]
> We should have a team learning session for how to use the PagerDuty interface. Could somebody record themselves going through the full PagerDuty workflow as part of these instructions? (maybe we could use something like https://loom.com/ ?)

Maybe a demo during the next monthly team meeting?

File: projects/managed-hubs/incidents.md
Comment (choldgraf) [reply projects/managed-hubs/incidents.md:136]: Nice! One question: what if we began by using PagerDuty for the incident itself, and GitHub to define the post-mortem content?

The reason I mention this is because (aside from the cost questions), it might solve two other problems:

1. Making the post-mortem easily accessible. If we used GitHub issues for this (as we do now), it is quite easy to search through all of our historical records of post-mortems.
2. Tying the post-mortem to the issues we create. The post-mortem almost always involves the creation of a bunch of new github issues etc for follow-up, and if we used GitHub for the post-mortem content itself, it would allow us to cross-reference to issues / PRs more easily.

If over time it seems like this doesn't work for us, we could always upgrade to a business plan.

File: projects/managed-hubs/incidents.md
Comment (choldgraf) [inline projects/managed-hubs/incidents.md:171]:
```suggestion

% TODO: Add example incident reports for reference when they exist
```

File: projects/managed-hubs/incidents.md
Comment (choldgraf) [inline projects/managed-hubs/incidents.md:105]:
```suggestion
```

File: projects/managed-hubs/incidents.md
Comment (choldgraf) [reply projects/managed-hubs/incidents.md:128]: This looks good to me, but I think an unanswered question is **when and why** would a person other than the incident commander be the one to fill this out? Should we encourage the engineer most-involved in resolving the incident to fix this? Should we try to rotate between members? Should we encourage the incident commander to do this most of the time but only ask somebody else if they really can't do it in a timely fashion?

File: projects/managed-hubs/incidents.md
Comment (sgibson91) [reply projects/managed-hubs/incidents.md:136]: > If we used GitHub issues for this (as we do now), it is quite easy to search through all of our historical records of post-mortems.

I would push back on this being "easy" a little bit. We do have to sift them out from all of the other issues going on in the infrastructure repo, and it's a very busy repo. We could transfer those issues to https://github.com/2i2c-org/incident-reports though.

Comment (choldgraf) [reply projects/managed-hubs/incidents.md:136]: When I do this I just filter by the Hub Incident label. At least this restricts to only the "incident" issues...agree it would be nicer if it were like a little Jupyter Book site or something.

- https://github.com/2i2c-org/infrastructure/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3A%22type%3A+Hub+Incident%22

**Comment by yuvipanda** [comment]
@choldgraf IMO, the postmortem process is one of the *main* reasons to use pagerduty, and I'd really like to try it out this way! In the future, I think automated alerts (and alerting / escalation) should also go through here I think

File: projects/managed-hubs/incidents.md
Comment (yuvipanda) [reply projects/managed-hubs/incidents.md:128]: @choldgraf I think our current experiences have been that all the 3 roles have been basically the same person, and I think our wide timezone spread has made it difficult for delegation to happen as well. I think this is an important question to answer, but heavily constrained by our timezone setup. Do you think this should block this PR?

File: projects/managed-hubs/incidents.md
Comment (yuvipanda) [reply projects/managed-hubs/incidents.md:136]: I agree with @sgibson91! IMO, it gets completely lost in GitHub and it's really difficult to find - especially as it is an issue comment that gets edited. It is also an incredibly manual process right now, and after trying out the process once in pagerduty I think using it is going to make postmortems *far more likely*. Also you can go to pagerduty and see a list of all incident reports - https://2i2c-org.pagerduty.com/postmortems.

In the future, I think we can use pagerduty for escalations and automated alerts as well. So regardless of what we do, I really want to move *away* from 'edit a github issue comment' as the model for how we do incident reports.

**Comment by choldgraf** [comment]
Sounds good - I defer to the @2i2c-org/tech-team's wishes on this one. If y'all think this is the right system then we should go for it.

**Comment by yuvipanda** [comment]
@choldgraf I've added a link to examples too

**Review by choldgraf — APPROVED** [review]
I think that we are all in agreement that this is a good direction to move towards. We've also already paid for a years-worth of accounts for PagerDuty. I think that we should merge this PR in and begin iterating from there.

**Comment by yuvipanda** [comment]
I agree @choldgraf! Merged!

**Pull request merged**
